Adding AU915 V1.0.2rB Support

ElectronicallyE opened the pull request:
Hi @terrillmoore,

This is follow-on from https://github.com/mcci-catena/arduino-lmic/issues/1010.

As mentioned in my issues post, I have created this branch with the changes needed to support AU915 V1.0.2rB. I have updated README.md to reflect this change.

The only other change I have made is the addition of LMIC_selectSubBand(1) in ttn-otaa.ino to align with the other examples.

I have tested this code using both The Things Network and Chirpstack and am confident it works.

Your feedback and willingness to merge into the "master" would be appreciated.

ElectronicallyE committed:
Modified files to support AU915 1.0.2rB

See https://github.com/mcci-catena/arduino-lmic/issues/1010 for more information.

ElectronicallyE committed:
Added LMIC_DEBUG_LEVEL + LMIC_LORAWAN_SPEC_VERSION

ElectronicallyE committed:
Added LMIC_selectSubBand(1) to align with other examples.

ElectronicallyE committed:
Reference to AU915 V1.0.2 added

ElectronicallyE committed:
Tidied up wording for AU915

ElectronicallyE committed:
Changed wording for AU915

ElectronicallyE committed:
Added #

Pranau-R commented:
Hi @ElectronicallyE ,

Thank you for bringing up this Dwell time issue to our attention.
We have reviewed your code changes and are trying to build and test it from the branch [Adding-AU915-1.0.2-Support](https://github.com/ElectronicallyE/arduino-lmic/tree/Adding-AU915-1.0.2-Support), but we were facing the following build issue:

```
Using previously compiled file: C:\Users\Pranau\AppData\Local\Temp\arduino\sketches\56F930A92C46CBD0986C6AB7E0C640A7\libraries\arduino-lmic\lmic\lmic_us915.c.o
c:\Users\Pranau\Documents\Arduino\libraries\arduino-lmic\src\lmic\lmic_au915.c:307:9: error: expected identifier or '(' before 'if'
         if (dwellDelay > globalDutyDelay) {
         ^~
c:\Users\Pranau\Documents\Arduino\libraries\arduino-lmic\src\lmic\lmic_au915.c:310:9: error: expected identifier or '(' before 'if'
         if (globalDutyDelay != 0) {
         ^~
c:\Users\Pranau\Documents\Arduino\libraries\arduino-lmic\src\lmic\lmic_au915.c:313:1: error: expected identifier or '(' before '}' token
 }
 ^
```

by removing the parentheses from the [line 304](https://github.com/ElectronicallyE/arduino-lmic/blob/Adding-AU915-1.0.2-Support/src/lmic/lmic_au915.c#L304) of [lmic_au915.c](https://github.com/ElectronicallyE/arduino-lmic/blob/Adding-AU915-1.0.2-Support/src/lmic/lmic_au915.c), this build issue can be fixed.

We will test the changes and let you know how it works; meanwhile, you can do the necessary code changes and update the pull request.

Thank you.

ElectronicallyE committed:
Update lmic_au915.c

Added { to line 302 to fix issue on line 304

ElectronicallyE committed:
DR0 JoinReq for 1.0.2 instead of DR2 for 1.0.3 if statements

ElectronicallyE commented:
Hi @Pranau-R,

This evening I have tested the code.

I downloaded my modified branch and removed the existing library I had installed.

The only changes I made in addition to my branch are:

**lmic_project_config.h**
I uncommented the following lines:
```
#define CFG_au915 1
#define CFG_sx1276_radio 1
#define LMIC_LORAWAN_SPEC_VERSION   LMIC_LORAWAN_SPEC_VERSION_1_0_2
```

ttn-otaa.ino
I added my `APPEUI` (LSB), `DEVEUI` (LSB) and `APPKEY` (MSB) from the TTN Console.
I changed `const unsigned TX_INTERVAL = 60;` to `const unsigned TX_INTERVAL = 20;` for testing.
I changed my "pin mapping" from:

```
const lmic_pinmap lmic_pins = {
     .nss = 6,
     .rxtx = LMIC_UNUSED_PIN,
     .rst = 5,
     .dio = {2, 3, 4},
};
```
to the following:
```
const lmic_pinmap lmic_pins = {
    .nss = 18,
    .rxtx = LMIC_UNUSED_PIN,
    .rst = 14,
    .dio = {26, 33, 32}
};
```

I uploaded the code and the board connected first time. I power cycled the board several times without issue.

Please note that I do have "Resets join nonces" enabled as there isn't any persistence (memory) written into the code.
<img width="231" alt="image" src="https://github.com/user-attachments/assets/4960cde5-f3e0-4add-a3d9-e7c1d38c3384" />

You will also see I have also modified lmic_bandplan_au915.h to set the JoinReq DR to DR0 for 1.0.2 instead of DR2 for 1.0.3 as per my comment back near the start of my issue post:
https://github.com/mcci-catena/arduino-lmic/issues/1010#issuecomment-2982694063

Pranau-R commented:
Hi @ElectronicallyE,

Thank you for sharing the detailed test results.
We have pulled your updated code and will begin testing shortly. We will share our feedback based on the results.
For reference, we're using the [Catena 4610](https://store.mcci.com/products/mcci-catena-4610-integrated-node-for-lorawan-technology) board in our test setup.
Thanks again!

ElectronicallyE commented:
Hi @Pranau-R,

Checked again this morning by reinstalling my branch, using the ttn-otaa.ino from scratch and it worked without an issue, sending the join request on DR0. Did you make sure that your version of the library included the changes I made to lmic_bandplan_au915.h (see https://github.com/mcci-catena/arduino-lmic/pull/1011/commits/6de5bee7045c2dac6f055651f24d93ff64ef805e)?

<img width="791" alt="image" src="https://github.com/user-attachments/assets/09246ac6-d502-4660-92f5-bff705b88aa7" />

ElectronicallyE commented:
Hi @Pranau-R,

Here are the "General Information" settings for the device I tested with the code:

<img width="354" height="371" alt="image" src="https://github.com/user-attachments/assets/63ca2972-331e-4dfd-a115-20e864a0579a" />

Pranau-R commented:
Hi @ElectronicallyE ,
Thank you for sharing the information.

As mentioned earlier, we used the [compliance-otaa-halconfig.ino](https://github.com/mcci-catena/arduino-lmic/blob/master/examples/compliance-otaa-halconfig/compliance-otaa-halconfig.ino) sketch to enable more detailed debug output during our testing.

The following changes we made to this sketch:
- added `#define ARDUINO_LMIC_CFG_SUBBAND 1`
- Added our end device `APPEUI `(LSB), `DEVEUI `(LSB), and `APPKEY `(MSB) from the TTN Console.
- Set const unsigned `TX_INTERVAL` = 20.

No other modifications were done.

However, we observed that the device still joins the network using **DR2 (SF10)**. Please find the attached screenshots from the TTN Console and Terminal for your reference:

<img width="1915" height="791" alt="image" src="https://github.com/user-attachments/assets/1b03762d-fa2b-4096-ae29-7a21e1666a60" />

<img width="1864" height="387" alt="image" src="https://github.com/user-attachments/assets/0d6aabeb-6aee-4f3e-bd7e-73d66d2f5381" />

**Just for testing**, we modified the following in the **lmic_bandplan_au915.h** file::

```
#if LMIC_LORAWAN_SPEC_VERSION <= LMIC_LORAWAN_SPEC_VERSION_1_0_2
  #define	LMICuslike_getJoin125kHzDR()	(LORAWAN_DR0)
#else
  #define	LMICuslike_getJoin125kHzDR()	(LORAWAN_DR0) // changed this from DR2 to DR0
#endif
```
and 

```
#if LMIC_LORAWAN_SPEC_VERSION <= LMIC_LORAWAN_SPEC_VERSION_1_0_2
  #define LMICbandplan_getInitialDrJoin() (LORAWAN_DR0)
#else
  #define LMICbandplan_getInitialDrJoin() (LORAWAN_DR0) // changed this from DR2 to DR0
#endif
```

No changes were made in the **lmic_project_config.h** file, except uncommenting the following relevant definitions:

```
// project-specific definitions
#define CFG_au915 1
#define CFG_sx1276_radio 1
#define LMIC_LORAWAN_SPEC_VERSION   LMIC_LORAWAN_SPEC_VERSION_1_0_2
```

After the above modifications, we tested the device again. This time, it successfully joined the network using **DR0 (SF12)**. Screenshots from TTN Console and Terminal are attached below for reference:

<img width="1914" height="794" alt="image" src="https://github.com/user-attachments/assets/b6f656dd-d9a6-424d-bf4c-59a8b2d6b8d8" />

<img width="1866" height="431" alt="image" src="https://github.com/user-attachments/assets/4534c36b-e9dc-4d88-908d-b02fa0315a8e" />

We have tested these changes several times, and every time it joined using **DR0 (SF12)**. Please let us know if there is anything we might have missed.

Thank you!

ElectronicallyE commented:
Hi @Pranau-R,

That's unusual behaviour. Hopefully that resolves issues on your end. I am using the Arduino IDE 2.3.6.

ElectronicallyE commented:
Glad to hear you have had success @Pranau-R. Have you been able to figure out when you faced this issue and I didn't?

Pranau-R commented:
Hi @ElectronicallyE ,

This is not an issue. In our BSP, we have defined the **ARDUINO_LMIC_PROJECT_CONFIG_H** macro to point to a different configuration file. Therefore, this is expected behavior and not related to any issue within the arduino-lmic library.

Thank you.

ElectronicallyE commented:
Great, so does this mean these changes can now be merged?

terrillmoore commented:
Sorry for the delay. I will rebase this on the v6 code and merge shortly. (Unless you'd prefer to do the rebase.)

ElectronicallyE commented:
Hi @terrillmoore, thanks for the update. It's been some time since I've looked into this, so to ensure compatibility, I think it's best if you perform the rebase.
